src/me.rs

```diff
@@ -259,35 +307,142 @@ pub fn motion_estimation(
   }
 }
 
-fn full_search(
-  x_lo: isize, x_hi: isize, y_lo: isize, y_hi: isize, blk_h: usize,
-  blk_w: usize, p_org: &Plane, p_ref: &Plane, best_mv: &mut MotionVector,
-  lowest_cost: &mut u32, po: &PlaneOffset, step: usize, bit_depth: usize,
-  lambda: u32, pmv: &[MotionVector; 2], allow_high_precision_mv: bool
-) {
-  for y in (y_lo..=y_hi).step_by(step) {
-    for x in (x_lo..=x_hi).step_by(step) {
-      let plane_org = p_org.slice(po);
-      let plane_ref = p_ref.slice(&PlaneOffset { x, y });
-
-      let sad = get_sad(&plane_org, &plane_ref, blk_h, blk_w, bit_depth);
-
-      let mv = MotionVector {
-        row: 8 * (y as i16 - po.y as i16),
-        col: 8 * (x as i16 - po.x as i16)
-      };
+fn get_best_predictor(fi: &FrameInvariants,
+  po: &PlaneOffset, p_org: &Plane, p_ref: &Plane,
+  predictors: &[MotionVector],
+  bit_depth: usize, pmv: &[MotionVector; 2], lambda: u32,
+  mvx_min: isize, mvx_max: isize, mvy_min: isize, mvy_max: isize,
+  blk_w: usize, blk_h: usize) -> (MotionVector, u32) {
+  let mut center_mv = MotionVector{row: 0, col: 0};
+  let mut center_mv_cost = std::u32::MAX;
+
+  for init_mv in predictors.iter() {
+    let cost = get_mv_rd_cost(
+      fi, po, p_org, p_ref, bit_depth,
+      pmv, lambda, mvx_min, mvx_max, mvy_min, mvy_max,
+      blk_w, blk_h, *init_mv);
+
+    if cost < center_mv_cost {
+      center_mv = *init_mv;
+      center_mv_cost = cost;
+    }
+  }
 
-      let rate1 = get_mv_rate(mv, pmv[0], allow_high_precision_mv);
-      let rate2 = get_mv_rate(mv, pmv[1], allow_high_precision_mv);
-      let rate = rate1.min(rate2 + 1);
-      let cost = 256 * sad + rate * lambda;
+  return (center_mv, center_mv_cost);
+}
 
-      if cost < *lowest_cost {
-        *lowest_cost = cost;
-        *best_mv = mv;
+fn diamond_me_search(
+  fi: &FrameInvariants,
+  po: &PlaneOffset, p_org: &Plane, p_ref: &Plane,
+  predictors: &[Vec<MotionVector>; 3],
```

Comment:
I had an array of three `Vec<MotionVector>` because this is a remaining of inconclusive tests that applied different processing in function of the type of predictor (previous frame blocks, neighbor blocks, median).
I'll remove that distinction and just pass a slice of `&[MotionVector]` instead.